fix(forms): Don't override form group's dirty state when disabling controls

martinsik opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[X] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

```
form = this.fb.group({ a: ['a'], b: ['b'] });
form.markAsDirty();
console.log(form.dirty); // true

form.controls['a'].disable();
console.log(form.dirty); // false
```

I can manually change `dirty` state of a `FormGroup` but when I disable a single control it changes `dirty` state of its parent.

Demo: https://stackblitz.com/edit/angular-8bab4v?file=src%2Fapp%2Fapp.component.ts

Issue Number: #23309

## What is the new behavior?

When calling `disable()` on `AbstractControl` it now checks whether its parent is `dirty` and if its `dirty` state corresponds to the current state of its children. If not it means I had to change it myself with `markAsDirty` so I won't override parent's `dirty` state.

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

This could be a BC but it's probably very unlikely that anybody was relying on this behavior since its counter-intuitive.

martinsik committed:
fix(forms): don't override form group's dirty state when disabling controls

Update packages/forms/src/model.ts

Co-Authored-By: martinsik <martin.sikora.ahoj@gmail.com>

kara commented on packages/forms/src/model.ts:
```diff
@@ -491,10 +494,13 @@ export abstract class AbstractControl {
     this._onDisabledChange.forEach((changeFn) => changeFn(false));
```
Seems like `enable()` needs to be updated with the check as well. Otherwise, you'll have the same problem of resetting dirtiness when you re-enable the control.

kara commented on packages/forms/test/form_control_spec.ts:
```diff
@@ -991,6 +991,34 @@ import {FormArray} from '@angular/forms/src/model';
+        expect(c.dirty).toBe(false);
+
+        c.disable();
+        expect(a.dirty).toBe(true);
```
Can we add the following to this test (as well as the other one)?

```ts
c.enable();
expect(a.dirty).toBe(true);
```

I think this assertion would fail as the code is currently written.

kara reviewed (changes requested):
Thanks for the bug fix! Some comments below.

martinsik commented:
@kara I finally managed to run Bazel so I'll have a look at this ASAP. I've actually completely forgotten what I did here because it's been 6 months now so I'll need to recall what's going on in this PR.

kara commented:
@martinsik No worries, sorry for the initial delay. Ping me whenever it's ready and I'll have better turnaround time this time :-)

martinsik commented:
@kara I updated this PR. You were right, it really didn't work after calling `enable()` on a form control but it should be fixed now. I also added the test case you suggested.

I think I messed the history in this PR when I unintentionally referenced this PR id instead of the issue id but it's showing just 3 commits here https://github.com/angular/angular/pull/24591/commits which is correct.

martinsik commented:
I don't know why CLA is failing. Maybe because I merged one suggestion as a comit here on GitHub.

kara commented on packages/forms/src/model.ts:
```diff
@@ -817,6 +828,16 @@ export abstract class AbstractControl {
+   *
+   * @internal
+   */
+  _parentMarkedDirty(onlySelf?: boolean): boolean {
```
Thanks for pulling this out into its own method. Can we make it private though?

kara reviewed (changes requested):
One small nit, then should be good to go.

Can you squash your commits together into one? I think the commit in the middle is causing our linter job to fail.

kara committed:
fixup! fix(forms): don't override form group's dirty state when disabling controls

kara commented:
@martinsik FYI, I squashed the commits and fixed the nit to move the PR along.  If Google internal tests pass, this should be good to go.

kara reviewed (approved):
LGTM
